src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+            {
+                var import = syntaxGenerator.NamespaceImportDeclaration(name);
+                if (withFormatterAnnotation)
+                {
+                    import = import.WithAdditionalAnnotations(Formatter.Annotation);
+                }
+                return import;
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = new List<ReferenceLocation>();
+                var refLocationsInOtherDocuments = new List<ReferenceLocation>();
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol)
+                        .SelectMany(refSymbol => refSymbol.Locations);
+
+                    foreach (var refLocation in refLocationsForSymbol)
+                    {
+                        if (refLocation.Document.Id == document.Id)
+                        {
+                            refLocationsInCurrentDocument.Add(refLocation);
+                        }
+                        else
+                        {
+                            Debug.Assert(!PathUtilities.PathsEqual(refLocation.Document.FilePath, document.FilePath));
+                            refLocationsInOtherDocuments.Add(refLocation);
+                        }
+                    }
+                }
+
+                document = await FixDeclarationDocumentAsync(document, refLocationsInCurrentDocument, cancellationToken)
+                    .ConfigureAwait(false);
+                solution = document.Project.Solution;
```

Comment:
nit: assigning over the passed in document/solution can often be confusing and can make things harder to debug.  For example, you acquired a semanticmodel above.  But now document/semanticmodel are not in sync anymore.  It's easy for people to then screw things up.  In general, provide a new name for these entities to make the code clearer.